text/0129-video-replay-envelope.md

```diff
@@ -0,0 +1,62 @@
+- Start Date: 2024-02-06
+- RFC Type: feature
+- RFC PR: [#129](https://github.com/getsentry/rfcs/pull/129)
+- RFC Status: draft
+
+# Summary
+
+In order to capture video replays and present them to the user, we need to define/decide:
+
+- how to transport the video data to the server
+- how to integrate the video data into the RRWeb JSON format
+- how to combine multiple replay chunks to a single replay session
+
+All of these influence one another and need to be considered together in a single RFC.
+
+Note: SDK-side implementation that is currently being worked on relies on taking screenshots and encoding them to a video.
+This is based on an evaluation where a video has much smaller size than a sequence of images (**TODO fix these: factor of X for 720p video**).
+
+# Motivation
+
+We need this to to capture replays on platforms where it's not possible/feasible to produce an HTML DOM (i.e. the native format supported by RRWeb). For example: mobile apps.
+
+<!-- # Supporting Data -->
+<!-- Metrics to help support your decision (if applicable). -->
+
+# Options Considered
+
+## Using a video, with EnvelopeItem:Attachment
+
+- From the SDK, we should be able to use the existing envelope item type, [`ReplayRecording`](https://github.com/getsentry/relay/blob/5fd3969e88d3eea1f2849e55b61678cac6b14e44/relay-server/src/envelope.rs#L115C5-L115C20) to transport the video data.
```

Comment:
The `ReplayRecording` envelope type should contain RRWeb formatted JSON.  It has the following components: a headers JSON object, a new line character, RRWeb recording JSON.

The headers object must contain a segment_id key but can be extended if your use case requires it:

```json
{"segment_id": 12}
```

If you're going to send binary video data on this event I would recommend extending the headers with a `type` key and a `version` key.  E.g. `type: video` and `version: 2`.  You will also need to track the offsets.  The RRWeb JSON and the video offsets.  E.g. `rrweb-offsets: [22, 3461]` and `video-offsets: [3462, 9555]`.

The video can be appended to the end of the file (immediately after the RRWeb JSON).

Reply:
Alternatively we can introduce a third envelope type and process the video independently of the RRWeb JSON.  This has pros and cons but it preserves backwards compatibility on the recording endpoint and doesn't overly complicate that consumer with a bunch of forking attributes (type, version).

Reply:
OK, I've changed the RFC to propose a new envelope item type